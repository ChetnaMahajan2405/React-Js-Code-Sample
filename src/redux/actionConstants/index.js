/**
 * User Constants
 */

// LOGIN
export const LOGIN_START = "LOGIN_START";
export const LOGIN_SUCCESS = "LOGIN_SUCCESS";
export const LOGIN_FAILURE = "LOGIN_FAILURE";


// SEARCH
export const SEARCH_START = "SEARCH_START";
export const SEARCH_SUCCESS = "SEARCH_SUCCESS";
export const SEARCH_FAILURE = "SEARCH_FAILURE";
export const SEARCH_CLEAR = "SEARCH_CLEAR";

// LOGOUT
export const LOGOUT = "LOGOUT";