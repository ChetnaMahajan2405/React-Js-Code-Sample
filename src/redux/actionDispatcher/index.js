/**
 * @function Calls the service and then dispatch action Object as a response to reducer.
 * @param service - Name of the Service that handles all the input parameters required to be pass in service.
 * @param actionTypeStart - type : action start
 * @param actionTypeSuccess - type : action Success
 * @param actionTypeFailure - type : action Failure
 * @returns {function()}
 * @constructor
 */

export default function DispatchActionToReducer(
  service,
  actionTypeStart,
  actionTypeSuccess,
  actionTypeFailure,
  callback
) {
  return dispatch => {
    dispatch({
      type: actionTypeStart
    });
    service
      .then(response => {
        if (response.results) {
          dispatch({
            type: actionTypeSuccess,
            data: response.results
          });
          if (callback) callback(null, response.results);
        } else {
          dispatch({
            type: actionTypeFailure,
            error: response
          });
        }
      })
      .catch(error => {
        dispatch({
          type: actionTypeFailure,
          error
        });
        if (callback) callback(error);
      });
  };
}

export function SimpleDispatchActionToReducer(type, data, callback) {
  return dispatch => {
    dispatch({ type, data });
    if (callback) callback();
  };
}
