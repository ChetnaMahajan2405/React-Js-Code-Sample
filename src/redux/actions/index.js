export {
  login,
  logout 
} from "./userActions";

export {
  search,
  clear
} from './searchActions';