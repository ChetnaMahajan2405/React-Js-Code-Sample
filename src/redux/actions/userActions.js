/**
 *  Imports
 */

import { sessionService } from "redux-react-session";
import {
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  LOGIN_START,
  LOGOUT
} from "../actionConstants";

import 
  dispatchActionToReducer,
  { SimpleDispatchActionToReducer }
from "../actionDispatcher";
import {
  login as loginService,
} from "../services";

/**
 *  Action Creators for user login
 */
export const login = (email, password, callback) => {
  return dispatchActionToReducer(
    loginService(email),
    LOGIN_START,
    LOGIN_SUCCESS,
    LOGIN_FAILURE,
    (error, res) => {

      if(res && res.length) {
        /**
         *  Creating session and saving user details on success
         */
        res.map(r => {
          if(r.birth_year.toLowerCase() === password && r.name.toLowerCase() === email) {
            return sessionService
              .saveSession({
                token: "token",
              })
              .then(() => {
                return sessionService.saveUser(r)
              })
              .then(() =>  callback(error, r));
          } 
          return callback({
            message: 'Invalid email or password'
          }); 
        })

      } else callback({
        message: 'Invalid email or password'
      }); 
    }
  );
};

/**
 *  Action Creators for user logout
 */
export const logout = callback => {
  return SimpleDispatchActionToReducer(LOGOUT, null, callback);
};

