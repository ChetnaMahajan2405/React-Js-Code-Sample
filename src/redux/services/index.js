// export API from "./serviceList";
export { default as apiCall } from "./service";
export {
  default as login
} from "./user";
export {
  default as search
} from "./search";