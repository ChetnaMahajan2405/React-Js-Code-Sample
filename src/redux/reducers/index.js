import { combineReducers } from "redux";
import { sessionReducer } from "redux-react-session";
import { reducer as toastrReducer } from "react-redux-toastr";

import { LOGOUT } from "../actionConstants";

import user from "./userReducer";
import search from "./searchReducer";

const appReducer = combineReducers({
  user,
  search,
  session: sessionReducer,
  toastr: toastrReducer
});

const rootReducer = (state, action) => {
  let newState = state;

  if (action.type === LOGOUT) {
    newState = {};
  }

  return appReducer(newState, action);
};

/** export this module */
export default rootReducer;
