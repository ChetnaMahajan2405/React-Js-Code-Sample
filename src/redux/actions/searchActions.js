/**
 *  Imports
 */
import {
  SEARCH_SUCCESS,
  SEARCH_FAILURE,
  SEARCH_START,
  SEARCH_CLEAR
} from '../actionConstants';
import dispatchActionToReducer, {
  SimpleDispatchActionToReducer
} from '../actionDispatcher';
import { search as searchService } from '../services';

/**
 *  Action Creators for planets search
 */
export const search = planet => {
  console.log(planet)
  return dispatchActionToReducer(
    searchService(planet),
    SEARCH_START,
    SEARCH_SUCCESS,
    SEARCH_FAILURE,
  );
};

/**
 *  Action Creators for clearing search results
 */
export const clear = callback => {
  return SimpleDispatchActionToReducer(SEARCH_CLEAR, null, callback);
};

