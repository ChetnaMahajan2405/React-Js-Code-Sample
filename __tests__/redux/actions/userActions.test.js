import configureStore from 'redux-mock-store';
import ReduxThunk from 'redux-thunk';

// Actions to be tested
import { logout, login } from '../../../src/redux/actions';

const mockStore = configureStore([ReduxThunk]);
const store = mockStore();

describe('selectAvatar', () => {
  it('Dispatches the correct action and payload', () => {
    const expectedActions = [
      {
        type: 'LOGOUT',
        data: null
      }
    ];
    store.dispatch(logout());
    expect(store.getActions()).toEqual(expectedActions);
  });

  it('should handle LOGIN_SUCCESS', async () => {
    const initialState = {
      login: {
        data: {},
        error: {}
      }
    };
    const store = mockStore(initialState);
    const response = '{"statusCode": "200", "results": [{"name": "Luke SkyWalker", "birth_year": "19BBY" }]}';
    await fetch.mockResponseSuccess(response)

    store.dispatch(login('Luke SkyWalker', '19BBY'));
    expect(store.getActions()).toMatchSnapshot();
        

  });


  it('should handle LOGIN_FAILURE', async () => {
    const initialState = {
      login: {
        data: {},
        error: {}
      }
    };
    const store = mockStore(initialState);
    const response = '{"statusCode": "400"}';
    await fetch.mockResponseFailure(response)
    store.dispatch(login('Luke SkyWalker', '19BBY'));
    expect(store.getActions()).toMatchSnapshot();


  });
});
