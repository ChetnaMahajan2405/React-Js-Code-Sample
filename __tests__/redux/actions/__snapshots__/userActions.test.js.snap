// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectAvatar should handle LOGIN_FAILURE 1`] = `
Array [
  Object {
    "type": "LOGIN_START",
  },
]
`;

exports[`selectAvatar should handle LOGIN_SUCCESS 1`] = `
Array [
  Object {
    "type": "LOGIN_START",
  },
]
`;
